saleor/payment/gateways/np_atobarai/api.py

```diff
@@ -0,0 +1,237 @@
+
+        return PaymentResult(
+            status=status,
+            psp_reference=transaction_id,
+            errors=error_messages,
+        )
+
+
+def cancel_transaction(
+    config: ApiConfig, payment_information: PaymentData
+) -> PaymentResult:
+    with np_atobarai_opentracing_trace("np-atobarai.checkout.payments.cancel"):
+        action = TRANSACTION_CANCELLATION
+        payment_id = payment_information.payment_id
+        payment = Payment.objects.filter(id=payment_id).first()
+
+        if not payment:
+            return error_payment_result(
+                add_action_to_code(action, error_code=PAYMENT_DOES_NOT_EXIST)
+            )
+
+        psp_reference = payment.psp_reference
+
+        if not psp_reference:
+            return error_payment_result(
+                add_action_to_code(action, error_code=NO_PSP_REFERENCE)
+            )
+
+        result, error_codes = cancel(config, psp_reference)
+
+        if error_codes:
+            error_messages = get_error_messages_from_codes(
+                action, error_codes=error_codes
+            )
+            return errors_payment_result(error_messages)
+
+        return PaymentResult(status=PaymentStatus.SUCCESS)
+
+
+def change_transaction(
+    config: ApiConfig,
+    payment: Payment,
+    payment_information: PaymentData,
+    refund_data: Optional[Dict[int, int]],
+) -> Optional[PaymentResult]:
+    with np_atobarai_opentracing_trace("np-atobarai.checkout.payments.change"):
+        if refund_data:
+            goods = get_refunded_goods(config, refund_data, payment_information)
+        else:
+            goods = get_goods_with_discount(config, payment_information)
+
+        data = {
+            "transactions": [
+                {
+                    "np_transaction_id": payment.psp_reference,
+                    "billed_amount": format_price(
+                        payment.captured_amount - payment_information.amount,
+                        payment_information.currency,
+                    ),
+                    "goods": goods,
+                }
+            ]
+        }
+
+        result, error_codes = np_request(
+            config, "patch", "/transactions/update", json=data
+        )
+
+        if not error_codes:
+            status = result["authori_result"]
+            transaction_id = result["np_transaction_id"]
+
+            if status == PaymentStatus.PENDING:
+                if cancel_error_codes := cancel(config, transaction_id).error_codes:
+                    handle_unrecoverable_state(
+                        payment.order, "cancel", transaction_id, cancel_error_codes
+                    )
+                error_messages = result["authori_hold"]
+                return errors_payment_result(error_messages)
+
+            return PaymentResult(
+                status=PaymentStatus.SUCCESS,
+            )
+
+        if PRE_FULFILLMENT_ERROR_CODE in error_codes:
+            return None
```

Comment:
Log here?

Reply:
Why should we log this result? It's not an exceptional flow. We're just asking here NP for transaction status. If it's after fulfilment report, we reregister transaction.

Reply:
It's silent `None` when we have actually `error_codes`

Reply:
Logging isn't for exceptional flow :o